Quadlet - add support for .image units (#19912)
containers/podman: docs/source/markdown/podman-systemd.unit.5.md

It is recommended to use a fully qualified image name rather than a short name, both for
performance and robustness reasons.

The format of the name is the same as when passed to `podman pull`. So, it supports using
`:tag` or digests to guarantee the specific image version.

As a special case, if the `name` of the image ends with `.image`, Quadlet will use the image
pulled by the corresponding `.image` file, and the generated systemd service contains
a dependency on the `$name-image.service`.

> Note that the corresponding `.image` file must exist.

Reviewer: That is _really_ _really_ _REALLY_ cool!

Author: Thanks

The format of the name is the same as when passed to `podman run`, so it supports e.g., using
`:tag` or using digests guarantee a specific image version.

As a special case, if the `name` of the image ends with `.image`, Quadlet will use the image
pulled by the corresponding `.image` file, and the generated systemd service contains
a dependency on the `$name-image.service`.

> Note that the corresponding `.image` file must exists.

Reviewer 2: ```suggestion
Note that the corresponding `.image` file must exist.
```

The path of a device which is mounted for the volume.

### `Driver=`

> Specify the volume driver name. When set to `image`, the `Image` key must also be set

Reviewer 2: ```suggestion
Specify the volume driver name. When set to `image`, the `Image` key must also be set.
```

> This is equivalent to the Podman `--driver` option

Reviewer 2: ```suggestion
This is equivalent to the Podman `--driver` option.
```

### `Group=`

The host (numeric) GID, or group name to use as the group for the volume

### `Image=`

> Specifies the image the volume is based on, when `Driver` is set the `image`.

Reviewer 2: I'm not real sure what you trying to say here, here's my best guess.  Or maybe "set to an `image`"

```suggestion
Specifies the image the volume is based on when `Driver` is set to the `image`.
```

It is recommended to use a fully qualified image name rather than a short name, both for
performance and robustness reasons.

> The format of the name is the same as when passed to `podman pull`, so it supports e.g., using

Reviewer 2: soft suggestion, I'd drop the "e.g.,"

Author: I've rephrased the sentence (which I actually copied, so it appears three times now)

> `:tag` or using digests guarantee a specific image version.

Reviewer 2: Soft suggestion
```suggestion
`:tag`.  Using digests guarantees a specific image version.
```

As a special case, if the `name` of the image ends with `.image`, Quadlet will use the image
pulled by the corresponding `.image` file, and the generated systemd service contains
a dependency on the `$name-image.service`.

> Note that the corresponding `.image` file must exists.

Reviewer 2: ```suggestion
Note that the corresponding `.image` file must exist.
```

The path of a device which is mounted for the volume.

### `Driver=`

Specify the volume driver name. When set to `image`, the `Image` key must also be set.

This is equivalent to the Podman `--driver` option.

### `Group=`

The host (numeric) GID, or group name to use as the group for the volume

### `Image=`

Specifies the image the volume is based on when `Driver` is set to the `image`.

> It is recommended to use a fully qualified image name rather than a short name, both for

Reviewer: Non-blocking nit: we could point to the `podman-pull` man page for details (as the log message does)

`systemd-%N` is used, which is the same as the unit name but with a `systemd-` prefix to avoid
conflicts with user-managed volumes.

## Image units [Image]

Image files are named with a `.image` extension and contain a section `[Image]` describing the
container image pull command. The generated service is a one-time command that ensures that the image
exists on the host, pulling it if needed.

Using image units allows containers and volumes to depend on images being automatically pulled. This is

> particularly interesting when using special options to control image pull.

Reviewer 2: ```suggestion
particularly interesting when using special options to control image pulls.
```
or perhaps "control image pull operations."

| AllTags=true                        | --all-tags                                      |
| Arch=aarch64                        | --arch=aarch64                                  |
| AuthFile=/etc/registry/auth\.json   | --authfile=/etc/registry/auth\.json             |
| CertDir=/etc/registery/certs        | --cert-dir=/etc/registery/certs                 |
| Creds=myname\:mypassword            | --creds=myname\:mypassword                      |
| DecryptionKey=/etc/registery\.key   | --decryption-key=/etc/registery\.key            |
| Image=quay.io/centos/centos\:latest | podman image pull quay.io/centos/centos\:latest |
| OS=windows                          | --os=windows                                    |
| PodmanArgs=--os=linux               | --os=linux                                      |
| Platform=linux/x86_64               | --platform=linux/x86_64                         |
| TLSVerify=false                     | --tls-verify=false                              |
| Variant=arm/v7                      | --variant=arm/v7                                |

### `AllTags=`

All tagged images in the repository are pulled.

> This is equivalent to the Podman `--all-tags` option

Reviewer 2: ```suggestion
This is equivalent to the Podman `--all-tags` option.
```

### `Arch=`

Override the architecture, defaults to hosts, of the image to be pulled.

> This is equivalent to the Podman `--arch` option

Reviewer 2: ```suggestion
This is equivalent to the Podman `--arch` option.
```

### `AuthFile=`

> Path  of  the  authentication file.

Reviewer: Here and below is see a number of double or triple whitespaces.

Author: Done

> This is equivalent to the Podman `--authfile` option

Reviewer 2: ```suggestion
This is equivalent to the Podman `--authfile` option.
```

### `CertDir=`

Use  certificates  at path (*.crt, *.cert, *.key) to connect to the registry.

> This is equivalent to the Podman `--cert-dir` option

Reviewer 2: ```suggestion
This is equivalent to the Podman `--cert-dir` option.
```

### `Creds=`

The `[username[:password]]` to use to authenticate with the registry, if required.

> This is equivalent to the Podman `--creds` option

Reviewer 2: ```suggestion
This is equivalent to the Podman `--creds` option.
```

### `DecryptionKey=`

The `[key[:passphrase]]` to be used for decryption of images.

> This is equivalent to the Podman `--decryption-key` option

Reviewer 2: ```suggestion
This is equivalent to the Podman `--decryption-key` option.
```

### `Image=`

The image to pull.
It is recommended to use a fully qualified image name rather than a short name, both for
performance and robustness reasons.

The format of the name is the same as when passed to `podman pull`, so it supports e.g., using

> `:tag` or using digests guarantee a specific image version.

Reviewer 2: if you change above, change here too

### `OS=`

Override the OS, defaults to hosts, of the image to be pulled.

> This is equivalent to the Podman `--os` option

Reviewer 2: ```suggestion
This is equivalent to the Podman `--os` option.
```

### `Platform=`

Specify  the  platform  for  selecting the image.

> This is equivalent to the Podman `--platform` option

Reviewer 2: ```suggestion
This is equivalent to the Podman `--platform` option.
```

> 

Reviewer 3: Is there any reason to expose all of --os, --arch and --variant when a single --platform option already covers all three.
We can keep quadlet simpler here as we do not need to worry about backwards compat. I think having one way to do the same thing is better.

Author: I see your point. However, two comments:
1. From what I can see in the code `-variant` is not covered by `--platform`. Maybe I'm reading it wrong
2. Using `--platform` requires the user to set the OS even if they only want to set the ARCH. So, if we decide not to support something in Quadlet, I opt for not supporting `platform`. WDYT?

Reviewer 3: That is likely incorrect docs, --platform supports a OS/ARCH/VARIANT tuple.
And the parser allows you to keep the OS empty, this works:
```
podman pull --platform /arm64/v8 alpine
```

I did not write any of this code and I don't the history of these options and maybe there is a subtle difference I am not aware of but guessing this is a mix of docker and backwards compat. I would assume @vrothberg knows more.

To me I like to keep simple and only offer one way to do the same thing. Of course that is personal preference and there is not much wrong with offering all options.

Author: Looking [here](https://github.com/containers/podman/blob/main/cmd/podman/images/pull.go#L144) I can see that `platform` gets parsed and the value is used for `os` and `arch`, but not `variant`. Having said that if the `arch` can be set to something like `arm64/v8` and will be treated correctly, then it should work. But, I don't know

Reviewer: --platform is supposed to include --os --arch and --variant.  If it doesn't please open  an issue.

I am personally in favor of having three fields: `OS`, `Architecture` and `Variant`.  It's much easier to brain parse and more expressive.

Author: @vrothberg I agree. So, I've removed the support for the `Platform` key 
As for its support for variant in platform. I'm not sure if there is an issue or not and I'm not sure how to test it.

### `PodmanArgs=`

This key contains a list of arguments passed directly to the end of the `podman image pull` command
in the generated file (right before the image name in the command line). It can be used to
access Podman features otherwise unsupported by the generator. Since the generator is unaware
of what unexpected interactions can be caused by these arguments, it is not recommended to use
this option.

The format of this is a space separated list of arguments, which can optionally be individually
escaped to allow inclusion of whitespace and other control characters.

This key can be listed multiple times.

### `TLSVerify=`

> Require HTTPS and verify certificates when contacting registries

Reviewer 2: ```suggestion
Require HTTPS and verification of certificates when contacting registries.
```

> This is equivalent to the Podman `--tls-verify` option

Reviewer 2: ```suggestion
This is equivalent to the Podman `--tls-verify` option.
```

### `Variant=`

Override the default architecture variant of the container image.

> This is equivalent to the Podman `--variant` option

Reviewer 2: ```suggestion
This is equivalent to the Podman `--variant` option.
```
